src/services/redisSessionStorage.ts

```diff
@@ -2,52 +2,128 @@
  * This service uses the Redis client to store and retrieve session information.
  */
 
-import { Either, left } from "fp-ts/lib/Either";
+import { Either, isLeft, left, right } from "fp-ts/lib/Either";
 import * as redis from "redis";
 import { User } from "../types/user";
 import { extractUserFromJson } from "../types/user";
 import { ISessionStorage } from "./iSessionStorage";
+import TokenService from "./tokenService";
+
+const SESSION_NOT_FOUND_MESSAGE = "Session not found";
+
+export interface ISessionState {
+  readonly expired: boolean;
+  readonly expireAt: string;
+  readonly newToken?: string;
+  readonly user?: User;
+}
 
 export default class RedisSessionStorage implements ISessionStorage {
   constructor(
     private readonly redisClient: redis.RedisClient,
-    private readonly tokenDuration: number
+    private readonly tokenDuration: number,
+    private readonly tokenService: TokenService
   ) {}
 
   /**
    * {@inheritDoc}
    */
   public set(token: string, user: User): void {
-    // Set key to hold the string value. This data is set to expire (EX) after
-    // `this.tokenDuration` seconds.
-    // @see https://redis.io/commands/set
-    this.redisClient.set(token, JSON.stringify(user), "EX", this.tokenDuration);
+    // Set a key to hold the fields value.
+    // @see https://redis.io/commands/hmset
+    this.redisClient.hmset(token, {
+      data: JSON.stringify(user),
+      timestamp: Date.now()
+    });
   }
 
   /**
    * {@inheritDoc}
    */
-  public get(token: string): Promise<Either<Error, User>> {
-    const client = this.redisClient;
-
+  public get(token: string): Promise<Either<Error, ISessionState>> {
     return new Promise(resolve => {
-      // Get the value of key.
-      // @see https://redis.io/commands/get
-      client.get(token, (err, value) => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
         if (err) {
-          resolve(left<Error, User>(new Error(err.message)));
+          resolve(left<Error, ISessionState>(new Error(err.message)));
         } else {
           if (value === null || value === undefined) {
-            resolve(left<Error, User>(new Error("Session not found")));
+            resolve(
+              left<Error, ISessionState>(new Error(SESSION_NOT_FOUND_MESSAGE))
+            );
           } else {
-            const errorOrUser = extractUserFromJson(value);
+            // check if the token has expired. We don't remove expired token
+            // because client can later refresh the session.
+            const timestamp = +value.timestamp;
+            if (timestamp + this.tokenDuration * 1000 < Date.now()) {
+              resolve(
+                left<Error, ISessionState>(new Error("Token has expired"))
```

Comment:
not sure why this error string is hardcoded here but the one on line 53 is defined as a constant (`SESSION_NOT_FOUND_MESSAGE`), either is fine, just stick to the same style for all

Reply:
this is because `Session not found` is repeated more than two times and `no-duplicate-string` complains about that. I'll move all errors in costants